hudi-io/src/main/java/org/apache/hudi/io/storage/HoodieStorage.java

```diff
@@ -0,0 +1,326 @@
+   *
+   * @param location given location.
+   * @return the statuses of the files/directories in the given location.
+   * @throws FileNotFoundException when the location does not exist.
+   * @throws IOException           IO error.
+   */
+  List<HoodieFileStatus> listDirectEntries(HoodieLocation location) throws IOException;
+
+  /**
+   * Lists the statuses of all files under the give location recursively.
+   *
+   * @param location given location.
+   * @return the statuses of the files under the given location.
+   * @throws FileNotFoundException when the location does not exist.
+   * @throws IOException           IO error.
+   */
+  List<HoodieFileStatus> listFiles(HoodieLocation location) throws IOException;
+
+  /**
+   * Lists the statuses of the direct files/directories in the given location
+   * and filters the results, if the path is a directory.
+   *
+   * @param location given location.
+   * @param filter   filter to apply.
+   * @return the statuses of the files/directories in the given location.
+   * @throws FileNotFoundException when the location does not exist.
+   * @throws IOException           IO error.
+   */
+  List<HoodieFileStatus> listDirectEntries(HoodieLocation location,
+                                           HoodieLocationFilter filter) throws IOException;
+
+  /**
+   * Returns all the files that match the locationPattern and are not checksum files,
+   * and filters the results.
+   *
+   * @param locationPattern given pattern.
+   * @param filter          filter to apply.
+   * @return the statuses of the files.
+   * @throws IOException IO error.
+   */
+  List<HoodieFileStatus> globEntries(HoodieLocation locationPattern, HoodieLocationFilter filter)
+      throws IOException;
+
+  /**
+   * Renames the location from old to new.
+   *
+   * @param oldLocation source location.
+   * @param newLocation destination location.
+   * @return {@true} if rename is successful.
+   * @throws IOException IO error.
+   */
+  boolean rename(HoodieLocation oldLocation, HoodieLocation newLocation) throws IOException;
+
+  /**
+   * Deletes a directory at location.
+   *
+   * @param location directory to delete.
+   * @return {@code true} if successful.
+   * @throws IOException IO error.
+   */
+  boolean deleteDirectory(HoodieLocation location) throws IOException;
+
+  /**
+   * Deletes a file at location.
+   *
+   * @param location file to delete.
+   * @return {@code true} if successful.
+   * @throws IOException IO error.
+   */
+  boolean deleteFile(HoodieLocation location) throws IOException;
+
+  /**
+   * Qualifies a path to one which uses this storage and, if relative, made absolute.
+   *
+   * @param location to qualify.
+   * @return Qualified location.
+   */
+  HoodieLocation makeQualified(HoodieLocation location);
+
+  /**
+   * @return the underlying file system instance if exists.
+   */
+  Object getFileSystem();
```

Comment:
Not sure if we should make the return type generic.

Reply:
After complete refactoring using HoodieStorage, this method should not be used anymore.  However, to make incremental progress on refactoring, in some places I have to get the internal FileSystem instance.  So for now, I think this is fine.